dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,720 @@
+        if isinstance(return_annotation, tuple):
+            return len(return_annotation)
+        elif return_annotation is Signature.empty:
+            return 0
+        else:
+            return 1
+
+    def _get_worker_id(self):
+        """
+        Getter for 'worker_id'. In case of torch data worker it is the worker info, and
+        in case of a thread the thread identifier
+        """
+        worker_info = _torchdata.get_worker_info()
+        return worker_info.id if worker_info else threading.get_ident()
+
+    def _get_worker_data(self):
+        if self._worker_data is None:
+            self._worker_data = _DALIProxy._WorkerData(self._get_worker_id())
+        return self._worker_data
+
+    def _add_sample(self, bound):
+        """
+        Adds a sample to the current batch. In the collate function, we mark the batch as
+        complete. When a completed batch is encountered, a new batch should be started.
+        """
+        worker_data = self._get_worker_data()
+        if worker_data.pipe_run_ref is None or worker_data.pipe_run_ref.is_complete:
+            worker_data.pipe_run_ref = DALIPipelineRunRef(
+                batch_id=(worker_data.worker_id, worker_data.next_worker_batch_idx)
+            )
+            worker_data.next_worker_batch_idx += 1
+            worker_data.batch_sample_idx = 0
+
+        for name, value in bound.arguments.items():
+            if name not in worker_data.pipe_run_ref.call_args:
+                worker_data.pipe_run_ref.call_args[name] = []
+            worker_data.pipe_run_ref.call_args[name].append(value)
+
+        ret = tuple(
+            DALIOutputSampleRef(
+                self,
+                pipe_run_ref=worker_data.pipe_run_ref,
+                output_idx=i,
+                sample_idx=worker_data.batch_sample_idx,
+            )
+            for i in range(self._num_outputs)
+        )
+        # unpack single element tuple
+        if len(ret) == 1:
+            ret = ret[0]
+        worker_data.batch_sample_idx += 1
+        return ret
+
+    def _schedule_batch(self, pipe_run_ref):
+        """
+        Schedules a batch for execution by appending it to the DALI input queue.
+        """
+        if not pipe_run_ref.call_args:
+            raise RuntimeError("No inputs for the pipeline to run (was it already scheduled?)")
+        if not pipe_run_ref.is_complete:
+            raise RuntimeError("Batch is not marked as complete")
+        _nvtx.range_push(f"dali_input_q put {pipe_run_ref.batch_id}")
+
+        dali_input_q_item = DALIPipelineRunRef(pipe_run_ref.batch_id)
+        dali_input_q_item.call_args = pipe_run_ref.call_args
+        dali_input_q_item.is_complete = True
+        dali_input_q_item.is_scheduled = True
+        self._dali_input_q.put(dali_input_q_item)
+
+        pipe_run_ref.call_args = {}
+        pipe_run_ref.is_scheduled = True
+        _nvtx.range_pop()
+        return pipe_run_ref
+
+    def __call__(self, *args, **kwargs):
```

Comment:
How can this possibly ever work when you overwrite `self.__call__` in `__init__`?